system/jlib/jtrace.cpp

```diff
@@ -69,3 +100,740 @@ const char* LogTrace::queryLocalId() const
+
+        return true;
+    }
+
+    opentelemetry::v1::trace::SpanContext querySpanContext() const
+    {
+        if (span != nullptr)
+            return span->GetContext();
+
+        return opentelemetry::trace::SpanContext::GetInvalid();
+    }
+
+    const char * queryTraceName() const
+    {
+        return tracerName.get();
+    }
+
+protected:
+    CSpan(const char * spanName, CSpan * parent)
+    {
+        name.set(spanName);
+        localParentSpan = parent;
+        if (localParentSpan != nullptr)
+            tracerName.set(parent->queryTraceName());
+        //type = spanType;
+
+       // init();
+    }
+
+    CSpan(const char * spanName, const char * nameOfTracer)
+    {
+        name.set(spanName);
+        localParentSpan = nullptr;
+        tracerName.set(nameOfTracer);
+        //type = spanType;
+        
+    }
+
+    void init()
+    {
+        auto provider = opentelemetry::trace::Provider::GetTracerProvider();
+
+        //what if tracerName is empty?
+        auto tracer = provider->GetTracer(tracerName.get());
+
+        if (localParentSpan != nullptr)
+            injectlocalParentSpan(localParentSpan);
+
+        span = tracer->StartSpan(name.get(), {}, opts);
+
+        if (span != nullptr)
+        {
+            storeSpanContext();
+
+            StringBuffer out;
+            toString(out);
+            DBGLOG("Span start: (%s)", out.str());
+        }
+    }
+
+    void storeSpanContext()
+    {
+        if (span != nullptr)
+        {
+            storeTraceID();
+            storeSpanID();
+            storeTraceFlags();
+        }
+    }
+
+    void injectlocalParentSpan(CSpan * localParentSpan)
+    {
+        if (localParentSpan != nullptr)
+        {
+            auto localParentSpanCtx = localParentSpan->querySpanContext();
+            if(localParentSpanCtx.IsValid())
+                opts.parent = localParentSpanCtx;
+        }
+    }
+
+    void storeTraceID()
+    {
+        traceID.clear();
+
+        if (!span)
+            return;
+
+        auto spanCtx = span->GetContext();
+        if (!spanCtx.IsValid())
+            return;
+
+        if (!spanCtx.trace_id().IsValid())
+            return;
+
+        char trace_id[32] = {0};
+        
+        spanCtx.trace_id().ToLowerBase16(trace_id);
+        traceID.set(trace_id, 32);
+    }
+
+    void storeSpanID()
+    {
+        spanID.clear();
+
+        if (!span)
+            return;
+
+        char span_id[16] = {0};
+        span->GetContext().span_id().ToLowerBase16(span_id);
+
+        spanID.set(span_id, 16);
+    }
+
+    void storeTraceFlags()
+    {
+        traceFlags.clear();
+
+        if (!span)
+            return;
+
+        char trace_flags[2] = {0};
+        span->GetContext().trace_flags().ToLowerBase16(trace_flags);
+
+        traceFlags.set(trace_flags, 2);
+    }
+
+    StringAttr name;
+    StringAttr tracerName;
+    StringAttr traceFlags;
```

Comment:
I went back and forth on this a few times, ultimately chose to clone to avoid a few lines in most situations.
Ultimately I should have lazy loaded the cloned values.